src/triton_fil/optional.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace triton { namespace backend { namespace fil {
+
+template <typename T>
+class optional {  // C++17: Switch to std::optional
+ public:
+  optional() : empty_{}, has_value_{false} {}
+
+  optional(const T& input_value) : value_{input_value}, has_value_{true} {}
+  optional(optional<T>&& other) : has_value_{other}
+  {
+    if (other) {
+      value_ = *other;
+    }
+  }
+
+  ~optional()
+  {
+    if (has_value_) {
+      value_.~T();
+    } else {
+      empty_.~empty_byte();
+    }
+  }
+
+  explicit operator bool() const { return has_value_; }
+  T& operator*() { return value_; }
+  T* operator->() { return &value_; }
+  optional& operator=(T&& new_value)
+  {
+    value_ = new_value;
+    has_value_ = true;
+    return *this;
+  }
+  optional& operator=(const T& new_value)
```

Comment:
It doesn't compile without that definition, but I'm scratching my head as to exactly why. The copy assignment operator is deleted because the type contains a union field, so maybe in that case it has to be explicitly defined? I'm honestly not sure, but we definitely need it.

Reply:
Hmmm, maybe in that case you could get away without the move assignment operator, and just keeping the copy assignment operator? r-value should also bind to a const l-value, and of the top of my head, I think that is also the binding precedence here. Although, if it doesn't work, feel free to just resolve this comment. Just trying to remove the duplicated code, that's all

Reply:
Aha! 
![image](https://user-images.githubusercontent.com/14876585/117768997-00ddd880-b1e8-11eb-8dc7-bcd513904ce2.png)

Reply:
So according to the chart, rvalue will bind to const&, meaning you can get away without needing a move assignment operator. However, totally forgot that this is a temporary workaround until C++17, so it's fine either way!